paddle/fluid/framework/details/inplace_op_pass.cc

```diff
@@ -83,490 +78,373 @@ const std::string kInplacedOpWhiteList[] = { // NOLINT
-                                 }
-                               });
-  return input_it == prev_op->inputs.end() ? nullptr : *input_it;
+std::vector<ir::Node *> *InplacePass::AllVersionVars(
+    const std::string &var_name) const {
+  auto iter = ssa_map_.find(var_name);
+  PADDLE_ENFORCE(iter != ssa_map_.end(), "cannot find var %s in ssa graph",
+                 var_name);
+  PADDLE_ENFORCE(!iter->second.empty(), "var %s is empty in ssa graph",
+                 var_name);
+  return &(iter->second);
 }
 
-InplacePass::InplacePass() : Pass() {
-  if (FLAGS_enable_inplace_whitelist) {
-    for (auto& s : kInplacedOpWhiteList) {
-      whitelist_.emplace(s);
+bool InplacePass::IsSkipVar(const std::string &var_name) const {
+  return skip_vars_.count(var_name) > 0;
+}
+
+bool InplacePass::IsLastVersionVar(ir::Node *var) const {
+  return AllVersionVars(var->Name())->back() == var;
+}
+
+bool InplacePass::CheckOpDeps(ir::Node *op,
+                              const std::vector<ir::Node *> &ops) const {
+  std::unordered_set<ir::Node *> other_ops(ops.begin(), ops.end());
+  other_ops.erase(op);
+  if (other_ops.empty()) return true;
+
+  // Traverse all preceding ops of op
+  std::queue<ir::Node *> queue;
+  std::unordered_set<ir::Node *> visited_ops;
+  queue.push(op);
+  visited_ops.insert(op);
+
+  // Visit all preceding ops of `op`, and erase it if any op is inside other_ops
+  // return true only if other_ops is empty(), which means that all `ops`
+  // depends on `op`.
+  while (!queue.empty()) {
+    auto *cur_op = queue.front();
+    queue.pop();
+
+    for (auto *in_var : cur_op->inputs) {
+      for (auto *in_op : in_var->inputs) {
+        if (visited_ops.count(in_op) != 0) {
+          continue;
+        }
+
+        visited_ops.insert(in_op);
+        queue.push(in_op);
+        other_ops.erase(in_op);
+        if (other_ops.empty()) return true;
+      }
     }
   }
+  return false;
 }
 
-void InplacePass::InitSSAGraphNodes() const {
-  std::unordered_map<std::string, std::unordered_set<ir::Node*>> all_vars;
-  for (auto* op : view_.AllOps()) {
-    for (auto* node : op->inputs) {
-      if (!node->IsVar() || node->IsCtrlVar()) continue;
-      if (all_vars[node->Name()].count(node) == 0) {
-        all_vars[node->Name()].emplace(node);
-        var_nodes_[node->Name()].emplace_back(node);
+void InplacePass::CollectSkipVars(ir::Graph *graph,
+                                  const std::vector<ir::Node *> &ops) const {
+  // 1. Collect op role vars
+  PADDLE_ENFORCE(graph->Has(details::kMemOptSkipVars),
+                 "Graph should have attr %s", details::kMemOptSkipVars);
+  auto &mem_opt_whitelist = graph->Get<MemOptSkipVars>(kMemOptSkipVars);
+  for (const auto &var : mem_opt_whitelist) {
+    skip_vars_.emplace(var);
+  }
+
+  // 2. track the nodes which used by parameter server.
+  // these node can not be inplaced, otherwise trainer
+  // pserver can not find each other name.
```

Comment:
Tiny Grammar issue: "each other name" should be "each other's name"

Reply:
Done.

Reply:
Done.